server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/ExportService.java

```diff
@@ -0,0 +1,128 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.webadmin.vault.routes;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.util.Collection;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStore;
+import org.apache.james.blob.export.api.BlobExportMechanism;
+import org.apache.james.core.MailAddress;
+import org.apache.james.core.User;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.DeletedMessageZipper;
+import org.apache.james.vault.search.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+class ExportService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
+
+    private final BlobExportMechanism blobExport;
+    private final BlobStore blobStore;
+    private final DeletedMessageZipper zipper;
+    private final DeletedMessageVault vault;
+
+    @Inject
+    @VisibleForTesting
+    ExportService(BlobExportMechanism blobExport, BlobStore blobStore, DeletedMessageZipper zipper, DeletedMessageVault vault) {
+        this.blobExport = blobExport;
+        this.blobStore = blobStore;
+        this.zipper = zipper;
+        this.vault = vault;
+    }
+
+    Mono<Void> export(User user, Query exportQuery, MailAddress exportToAddress,
+                      Runnable messageToExportCallback) {
+
+        return matchingMessages(user, exportQuery)
+            .doOnNext(any -> messageToExportCallback.run())
+            .collect(Guavate.toImmutableList())
+            .map(Collection::stream)
+            .map(sneakyThrow(messages -> zipData(user, messages)))
+            .flatMap(sneakyThrow(zippedStream -> blobStore.save(zippedStream, zippedStream.available())))
```

Comment:
blobStore.save() require to pass contentLength parameter, so I came up with inputStream.avaliable(). I not sure it's good, but it passes all tests

Reply:
Arg. 

```
     * Returns an estimate of the number of bytes that can be read (or
     * skipped over) from this input stream without blocking by the next
     * invocation of a method for this input stream.
```

Not want you think it is! It's buggy and dangerous - don't use it!

Reply:
The inputStream is `PipedInputStream`. I went to it's impl, seem like we can trust this:
```java
    /**
     * Returns the number of bytes that can be read from this input
     * stream without blocking.
     *
     * @return the number of bytes that can be read from this input stream
     *         without blocking, or {@code 0} if this input stream has been
     *         closed by invoking its {@link #close()} method, or if the pipe
     *         is {@link #connect(java.io.PipedOutputStream) unconnected}, or
     *          <a href="#BROKEN"> <code>broken</code></a>.
     *
     * @exception  IOException  if an I/O error occurs.
     * @since   JDK1.0.2
     */
    public synchronized int available() throws IOException {
        if(in < 0)
            return 0;
        else if(in == out)
            return buffer.length;
        else if (in > out)
            return in - out;
        else
            return in + buffer.length - out;
    }
```

Reply:
I think that it does not compute what you think `the number of bytes that can be read from this input stream without blocking` means that there can be more bytes but I will have to fill the buffer with a blocking call. Is it the information you are looking for?

Reply:
we discussed this during the daily, and we will modify the blob api to not require the size

Reply:
Please @trantienduchn implement that with a copy to a local file - we have no better alternative!

Let's kill that beautiful piped input stream....